src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudService.kt

```diff
@@ -123,64 +139,91 @@ class CosmasGoogleCloudService(private val freeBucketName: String,
+
+                        val newText = PatchCorrector.applyPatch(patches, text)
+                        val cosmasHash = md5Hash(newText)
+                        if (patches.last().actualHash == cosmasHash) {
+                            commitFromMemoryToGCS(request.info, fileId, newText)
+                            LOG.info("File={} has been committed", fileId)
+                        } else {
+                            val actualHash = patches.last().actualHash
+                            LOG.error("Commit failure: File={} has Cosmas hash={}, but last actual hash={}",
+                                    fileId, cosmasHash, actualHash)
                             val badFile = CosmasProto.FileInfo.newBuilder()
                                     .setFileId(fileId)
                                     .setProjectId(request.info.projectId)
                                     .build()
                             response.addBadFiles(badFile)
                         }
-                        else -> throw e
-                    }
+                    } catch (e: Throwable) {
+                        LOG.error("Error while applying patches to file={}", fileId)
+                        when (e) {
+                            is PatchCorrector.ApplyPatchException, is IllegalArgumentException -> {
+                                val badFile = CosmasProto.FileInfo.newBuilder()
+                                        .setFileId(fileId)
+                                        .setProjectId(request.info.projectId)
+                                        .build()
+                                response.addBadFiles(badFile)
+                            }
+                            else -> throw e
+                        }
 
+                    }
                 }
+            } catch (e: StorageException) {
+                handleStorageException(e, responseObserver)
+                return
             }
-        } catch (e: StorageException) {
-            handleStorageException(e, responseObserver)
-            return
         }
-
         responseObserver.onNext(response.build())
         responseObserver.onCompleted()
     }
 
+    // We have to pass newText there cause we can get new text version from this method
+    // only by applying patches to version in memory. In "commit" we re applying patches
+    // before we re calling this method, so we dont want to do it again.
+    // In "forcedCommit" new text version is given in request.
+    private fun commitFromMemoryToGCS(projectInfo: ProjectInfo, fileId: String, newText: String) {
+        val project = this.fileBuffer[projectInfo.projectId] ?: return
+        val fileVersion = project[fileId] ?: return
+
+        // Just committing buffered version with new content and current timestamp to GCS
+        val curTime = clock.millis()
+        val newVersion = fileVersion.toBuilder()
+                .setContent(ByteString.copyFrom(newText.toByteArray()))
```

Comment:
btw there is `ByteString.copeFromUtf8`: https://developers.google.com/protocol-buffers/docs/reference/java/com/google/protobuf/ByteString#copyFromUtf8-java.lang.String-

Reply:
done